feat(nuxt): Use --import as the default installation method (#12070)
getsentry/sentry-docs: docs/platforms/javascript/guides/nuxt/install/dynamic-import.mdx

---
title: Dynamic Import (experimental)
sidebar_order: 3
description: "Learn about how the Nuxt SDK leverages dynamic input() in the build output."
---

## Understanding the `import()` expression

The `import()` expression (also called "dynamic import") allows conditional and flexible module loading in ESM.
For the Sentry Nuxt SDK, it provides an approach to initialize server-side configuration before application startup.

The Sentry Nuxt SDK will be initialized before any other code runs and the Nitro server runtime code will be loaded later because of `import()`ing it.

> This early initialization is required to set up the SDK's instrumentation of various libraries correctly.

Author: I know the text might be a bit complicated, but I think the suggestion is missing some crucial parts. It makes the impression that Nitro is initialized early anyways. It's important to point out that the SDK adds this dynamic import to create this late initialization.

> This early initialization is required to set up the SDK's instrumentation of various libraries correctly.

Reviewer: +1 to the suggestion missing info.

> This early initialization is required to set up the SDK's instrumentation of various libraries correctly.

Reviewer 2: Ok, that makes sense. I think in that case, it's fine to keep this as is :)

## Initializing Sentry with Dynamic `import()`

Enable the dynamic `import()` by setting `autoInjectServerSentry`:

```typescript {filename:nuxt.config.ts} {3}
export default defineNuxtConfig({
  sentry: {
    autoInjectServerSentry: 'experimental_dynamic-import'
  },
})
```

> After setting this, the Sentry Nuxt SDK adds a Rollup plugin which wraps the server entry file with a dynamic `import()`.

Reviewer: l: This feels like a bit too much detail. Even the concept of `server entry file` is confusing, it sounds a bit like our server config file.

Maybe just saying that the build system will take care of ensuring import order using dynamic imports?

## Scenarios where `import()` doesn't work

Depending on your setup and which version of Nitro is used, the server-side is sometimes not correctly initialized.

> The build output **must not** include a regular `import` of the Nitro runtime (e.g. `import './chunks/nitro/nitro.mjs'`).

Reviewer: m: I think we need a simpler way of telling people how to detect this other than looking through some build files they might not even have access to.

How about monitoring startup logs and detecting whether otel wrapped http instead?

I think the technical details are better kept in the github issue linked below.

Author: https://github.com/getsentry/sentry-javascript/issues/14638

I added an issue for this and will update the docs after adding this warning 👍

}
```

> You can also check out the guide for using the <PlatformLink to="/install/cli-import">CLI flag `--import`</PlatformLink> or <PlatformLink to="/install/top-level-import/">a top-level import</PlatformLink> as this might be a better choice for you.

Author: ```suggestion
You can also check out the guide for installing the SDK with a <PlatformLink to="/install/cli-import">CLI flag `--import`</PlatformLink> or a <PlatformLink to="/install/top-level-import/">top-level import</PlatformLink>.
```
What do you think of this?
